### DISCUSSION TITLE
fs doesn't exist

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1220] [Q&A]
when i intergate langium with glsp,vscode-extension an error like below
![image](https://github.com/eclipse-langium/langium/assets/26995040/74dfbb9b-bbaf-405b-a279-ad459cfafb10)
how to solve it ,thanks a lot

**Comment by msujew** [comment]
@zrwsmd Can you share your bundler config? It looks like you're trying to bundle for a browser platform but your language server code import from `langium/node`. 

You should either bundle using node as a target or use the `EmptyFileSystem` instead of the import from `langium/node`.

**Reply by zrwsmd** [reply]
which file is  bundler config

**Reply by msujew** [reply]
If you've used the yeoman generator to create your project, you should have an `esbuild.mjs` file in your project.

**Reply by zrwsmd** [reply]
i am afraid i did not have this file , i use glsp-vscode-integration-master as my  basic framework,then i add langium server into it.
    my requiremnet is when user performs a change in the diagram(my requirement diagram is industrial Programming Language LD(Ladder Diagram,(right part))) ,corresponding lauguage like C++ or PLC ST (Structured Text,left part）language generated
![image](https://github.com/eclipse-langium/langium/assets/26995040/16f65fdb-5147-4c91-bd81-7f33edc7649e)

**Reply by msujew** [reply]
I see. I assume they're using webpack, in that case the bundler file should be a `webpack.config.js` file.

**Reply by zrwsmd** [reply]
below is webpack.config.js file

<details>

```js
//@ts-check

'use strict';

const path = require('path');
const webpack = require('webpack');

/**@type {import('webpack').Configuration}*/
const config = {
    target: 'webworker', // vscode extensions run in webworker context for VS Code web -> https://webpack.js.org/configuration/target/#target
    //entry: './src/workflow-extension.ts',
    entry: {
        // the entry point of this extension,  -> https://webpack.js.org/configuration/entry-context/
        extension: './src/web-extension.ts',
        server: { import: './src/main.ts', filename: 'server-main.js' }
     },
    output: {
        // the bundle is stored in the 'dist' folder (check package.json),  -> https://webpack.js.org/configuration/output/
        path: path.resolve(__dirname, 'dist'),
        filename: 'extension.js',
        libraryTarget: 'commonjs2'
    },
    mode: 'development',
    devtool: 'eval-source-map',
    externals: {
        vscode: 'commonjs vscode' // the vscode-module is created on-the-fly and must be excluded. Add other modules that cannot be webpack'ed,  -> https://webpack.js.org/configuration/externals/
    },
    resolve: {
        // support reading TypeScript and JavaScript files,  -> https://github.com/TypeStrong/ts-loader
        mainFields: ['browser', 'module', 'main'], // look for `browser` entry point in imported node modules
        extensions: ['.ts', '.js']
    },
    module: {
        rules: [
            {
                test: /\.ts$/,
                exclude: /node_modules/,
                use: [
                    {
                        loader: 'ts-loader'
                    }
                ]
            }
        ]
    },
    plugins: [
        new webpack.ProvidePlugin({
            process: 'process/browser' // provide a shim for the global `process` variable
        })
    ],
    performance: {
        hints: false
    }
};
module.exports = config;
```

</details>

**Reply by msujew** [reply]
Ok, so the project you're using as a template is building a vscode web extension. These run in the browser/webworker context and not in node - which is the reason you're getting this error in the first place. `fs` is a node.js dependency that cannot be used in the browser.

Unfortunately, I'm not sure how to help you there. I'm quite unfamiliar with GLSP. You're probably better off asking the people over at [the eclipse-glsp project](https://github.com/eclipse-glsp/glsp/discussions), since they're also quite familiar with language servers (and Langium is just a framework for building such).

**Comment by zrwsmd** [comment]
![image](https://github.com/eclipse-langium/langium/assets/26995040/be6cadbb-f63f-47b4-bc65-00e32b2d1ddc)

**Comment by zrwsmd** [comment]
web-extension.ts is same as glsp-vscode-integration-master/example/workflow/web-extension/src/workflow-extension.ts

**Comment by zrwsmd** [comment]
actually i am a little confused about framework,i am considering use glsp server,langium server ,vscode extension,vscode webview to achieve my requirement,but for project langium-main/examples/statemachine/src/extension.ts example ,it seems to launch a LanguageClient,whereas for glsp-vscode-integration-master/example/workflow/web-extension/src/workflow-extension.ts it code is totally different with  langium-main/examples/statemachine/src/extension.ts,for my requirement,should i also launch LanguageClient
